src/parcsr_mv/par_csr_filter_device.c

```diff
@@ -0,0 +1,356 @@
+                                         HYPRE_Complex    *A_offd_a,
+                                         HYPRE_BigInt     *A_col_map_offd,
+                                         HYPRE_Int        *B_diag_i,
+                                         HYPRE_Int        *B_diag_j,
+                                         HYPRE_Complex    *B_diag_a,
+                                         HYPRE_Int        *B_offd_i,
+                                         HYPRE_Int        *B_offd_j,
+                                         HYPRE_Complex    *B_offd_a,
+                                         HYPRE_Int        *col_map_marker)
+{
+   const HYPRE_Int  row = hypre_gpu_get_grid_warp_id<1, 1>(item);
+   const HYPRE_Int  lane = hypre_gpu_get_lane_id<1>(item);
+   HYPRE_Int        p, q, pA, qA;
+   HYPRE_BigInt     big_block_size = (HYPRE_BigInt) block_size;
+
+   if (row >= num_rows)
+   {
+      return;
+   }
+
+   if (lane < 2)
+   {
+      p = read_only_load(A_diag_i + row + lane);
+      q = read_only_load(A_offd_i + row + lane);
+   }
+   pA = warp_shuffle_sync(item, HYPRE_WARP_FULL_MASK, p, 1);
+   p  = warp_shuffle_sync(item, HYPRE_WARP_FULL_MASK, p, 0);
+   qA = warp_shuffle_sync(item, HYPRE_WARP_FULL_MASK, q, 1);
+   q  = warp_shuffle_sync(item, HYPRE_WARP_FULL_MASK, q, 0);
+
+   HYPRE_Int diag_offset = B_diag_i[row];
+   for (HYPRE_Int j = p + lane;
+        warp_any_sync(item, HYPRE_WARP_FULL_MASK, j < pA);
+        j += HYPRE_WARP_SIZE)
+   {
+      const HYPRE_Int col     = (j < pA) ? read_only_load(A_diag_j + j) : 0;
+      HYPRE_Int       write   = (j < pA && (col % block_size) == (row % block_size));
+      hypre_mask      ballot  = hypre_ballot_sync(HYPRE_WARP_FULL_MASK, write);
+      HYPRE_Int       laneoff = hypre_popc(ballot & ((hypre_mask_one << lane) - 1));
+
+      if (write)
+      {
+         HYPRE_Int idx = diag_offset + laneoff;
+         B_diag_j[idx] = col;
+         B_diag_a[idx] = A_diag_a[j];
+      }
+
+      diag_offset += hypre_popc(ballot);
+   }
+
+   if (col_map_marker)
+   {
+      HYPRE_Int offd_offset = B_offd_i[row];
+      for (HYPRE_Int j = q + lane;
+           warp_any_sync(item, HYPRE_WARP_FULL_MASK, j < qA);
+           j += HYPRE_WARP_SIZE)
+      {
+         const HYPRE_Int    col        = (j < qA) ? read_only_load(A_offd_j + j) : 0;
+         const HYPRE_BigInt global_col = (j < qA) ? read_only_load(A_col_map_offd + col) : 0;
+         HYPRE_Int          write      = (j < qA) &&
+                                         (HYPRE_Int) (global_col % big_block_size) == (row % block_size);
+         hypre_mask         ballot     = hypre_ballot_sync(HYPRE_WARP_FULL_MASK, write);
+         HYPRE_Int          laneoff    = hypre_popc(ballot & ((hypre_mask_one << lane) - 1));
+
+         if (write)
+         {
+            HYPRE_Int idx = offd_offset + laneoff;
+            B_offd_j[idx] = col;
+            B_offd_a[idx] = A_offd_a[j];
+
+#ifndef HYPRE_USING_SYCL
```

Comment:
a minor thing: maybe `#if !defined`. Is it an SSDK requirement?

Reply:
It doesn't seem to be an xSDK requirement, but I can make the change

Reply:
If not, no need to change. but I do remember we talked about `#if defined()` is preferred over`#ifdef` by some software rule. I don't know why though.